mmdet/models/necks/bifpn.py

```diff
@@ -0,0 +1,326 @@
+
+    def __init__(self, input_channels, num_backbone_features, num_outs,
+                 channels, weight_method, act_fn, separable_conv, epsilon,
+                 input_offsets, feature_reduction, reduction):
+        super().__init__()
+
+        weight_method_list = ["attn", "fast_attn", "sum"]
+        self.num_outs = num_outs
+
+        assert self.num_outs >= 3
+        assert weight_method in weight_method_list
+
+        self.nodes = nn.ModuleList()
+
+        for i in range(num_outs - 1):
+            self.nodes.append(BiFPNNode(input_channels[num_outs - i - 2],
+                                        channels,
+                                        weight_method,
+                                        act_fn,
+                                        separable_conv,
+                                        epsilon,
+                                        input_offsets[i],
+                                        reduction[i],
+                                        feature_reduction,
+                                        reduction))
+
+        for i in range(1, num_outs):
+            self.nodes.append(BiFPNNode(input_channels[i],
+                                        channels,
+                                        weight_method,
+                                        act_fn,
+                                        separable_conv,
+                                        epsilon,
+                                        input_offsets[num_outs + i - 2],
+                                        reduction[num_outs + i - 2],
+                                        feature_reduction,
+                                        reduction))
+
+    def forward(self, inputs):
+        output = list(inputs)
+
+        for node in self.nodes:
+            output.append(node(output))
+
+        return tuple(output[-self.num_outs::])
+
+
+@NECKS.register_module
+class BiFPN(nn.Module):
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 input_indices,
+                 num_layers,
+                 num_outs,
+                 weight_method="fast_attn",
+                 act_fn="silu",
+                 separable_conv=True,
+                 epsilon=0.0001,
+                 base_reduction=8,
+                 reduction_ratio = 2.0):
+        super(BiFPN, self).__init__()
+        assert isinstance(in_channels, list)
+        self.num_backbone_features = len(in_channels)
+
+        assert self.num_backbone_features >= 2
+        assert num_outs - self.num_backbone_features >= 2
+
+        self.in_channels = in_channels
+        self.num_outs = num_outs
+
+        self.extra_convs = nn.ModuleList()
+
+        max_level = num_outs + self.num_backbone_features - 1
+        half_level = max_level // 2
+
+        input_offsets = []
+        feature_reduction = []
+        reduction = []
+
+        for i in input_indices:
+            feature_reduction.append(1 << i)
```

Comment:
I would not be so sure that it is true for all backbones.

Reply:
Tested multiple models by printing out [feature info](https://github.com/rwightman/efficientdet-pytorch/blob/master/effdet/efficientdet.py#L543) for [multiple models](https://github.com/rwightman/efficientdet-pytorch/blob/9370d27db522572e7c6364644146f88a8a9d2ea9/effdet/config/model_config.py#L81) from [this repo](https://github.com/rwightman/efficientdet-pytorch/blob/9370d27db522572e7c6364644146f88a8a9d2ea9/effdet/config/model_config.py#L81), and rule seems to work

Reply:
The problem is following: your implementation is not generalized, it won't work with all backbones.

See 'out_indices' and 'strides' here https://github.com/openvinotoolkit/training_extensions/blob/develop/pytorch_toolkit/object_detection/model_templates/face-detection/face-detection-0206/model.py

and here https://github.com/openvinotoolkit/training_extensions/blob/develop/pytorch_toolkit/object_detection/model_templates/face-detection/face-detection-0200/model.py

To make your bifpn more general, please introduce 'strides' parameter